### ISSUE TITLE
Notes are lost when updating the Standard Notes snap

### ISSUE THREAD

**Issue by pauldoo** [standardnotes/desktop#683]
I am using the snap version of Standard Notes on Ubuntu 20.04, though I also experienced this problem on an 18.04 machine.

When the version is updated all notes are lost and must be re-imported from a backup.  I am not using an account with Standard Notes.  The problem seems to be that the snap is storing the database in the revision specific folder (`~/snap/standard-nodes/<revision>/`), and not the common folder that is preserved during updates (`~/snap/standard-notes/common/`).

I first reported this problem on 2021-03-03 when 3.6 first released ([reddit thread](https://snapcraft.io/standard-notes)).  I am reporting it again now as the same issue occurred when 3.6.3 was published to snap.  I was asked on that comment thread to provide the contents of `~/.config/<app name>/log.log`, but that file does not exist on my system.

I would like to ask that:
1. As a short-term mitigation, please start publishing more than one version to snap using the channels feature.  This will allow users who experience problems with an upgrade to switch back to an older version.
2. Fix the snap version of Standard Notes so that the local database isn't lost on every snap update.

Here are some directory listings to illustrate the issue:

```
paul@box ~/snap/standard-notes> ls -lart
total 20
drwxr-xr-x  3 paul paul 4096 Mar  3 13:54 common/
drwxr-xr-x 14 paul paul 4096 Mar  3 13:56 33/
drwxr-xr-x  7 paul paul 4096 Mar  5 08:07 ../
lrwxrwxrwx  1 paul paul    2 Mar 26 07:50 current -> 34/
drwxr-xr-x  5 paul paul 4096 Mar 26 07:50 ./
drwxr-xr-x 14 paul paul 4096 Mar 26 07:50 34/
paul@box ~/snap/standard-notes> ls -la common 33 34
33:
total 60
drwxr-xr-x 14 paul paul 4096 Mar  3 13:56  ./
drwxr-xr-x  5 paul paul 4096 Mar 26 07:50  ../
drwxr-xr-x  7 paul paul 4096 Mar  3 13:54  .config/
drwxr-xr-x  2 paul paul 4096 Mar  3 13:54  Desktop/
drwxr-xr-x  2 paul paul 4096 Mar  3 13:54  Documents/
drwxr-xr-x  2 paul paul 4096 Mar  3 13:54  Downloads/
-rw-r--r--  1 paul paul   30 Mar  3 13:54  .last_revision
drwxr-xr-x  3 paul paul 4096 Mar  3 13:54  .local/
drwxr-xr-x  2 paul paul 4096 Mar  3 13:54  Music/
drwxr-xr-x  2 paul paul 4096 Mar  3 13:54  Pictures/
drwx------  3 paul paul 4096 Mar  3 13:54  .pki/
drwxr-xr-x  2 paul paul 4096 Mar  3 13:54  Public/
drwxr-xr-x  2 paul paul 4096 Mar  3 13:56 'Standard Notes Backups'/
drwxr-xr-x  2 paul paul 4096 Mar  3 13:54  Templates/
lrwxrwxrwx  1 paul paul   39 Mar  3 13:54  .themes -> /snap/standard-notes/33/data-dir/themes/
drwxr-xr-x  2 paul paul 4096 Mar  3 13:54  Videos/

34:
total 60
drwxr-xr-x 14 paul paul 4096 Mar 26 07:50  ./
drwxr-xr-x  5 paul paul 4096 Mar 26 07:50  ../
drwxr-xr-x  7 paul paul 4096 Mar 26 07:50  .config/
drwxr-xr-x  2 paul paul 4096 Mar 26 07:50  Desktop/
drwxr-xr-x  2 paul paul 4096 Mar 26 07:50  Documents/
drwxr-xr-x  2 paul paul 4096 Mar 26 07:50  Downloads/
-rw-r--r--  1 paul paul   30 Mar 26 07:50  .last_revision
drwxr-xr-x  3 paul paul 4096 Mar 26 07:50  .local/
drwxr-xr-x  2 paul paul 4096 Mar 26 07:50  Music/
drwxr-xr-x  2 paul paul 4096 Mar 26 07:50  Pictures/
drwx------  3 paul paul 4096 Mar 26 07:50  .pki/
drwxr-xr-x  2 paul paul 4096 Mar 26 07:50  Public/
drwxr-xr-x  2 paul paul 4096 Mar 26 07:58 'Standard Notes Backups'/
drwxr-xr-x  2 paul paul 4096 Mar 26 07:50  Templates/
lrwxrwxrwx  1 paul paul   39 Mar 26 07:50  .themes -> /snap/standard-notes/34/data-dir/themes/
drwxr-xr-x  2 paul paul 4096 Mar 26 07:50  Videos/

common:
total 12
drwxr-xr-x 3 paul paul 4096 Mar  3 13:54 ./
drwxr-xr-x 5 paul paul 4096 Mar 26 07:50 ../
drwxr-xr-x 6 paul paul 4096 Mar 26 07:50 .cache/
```

**Comment by pauldoo** [comment]
Extra information..
I had the [Parallel Installs](https://snapcraft.io/docs/parallel-installs) feature enabled in snap.  The documentation for that feature says that some snap directory locations are impacted, and applications need to be aware of that.

To see if this was part of the problem I: removed standard notes, disabled that feature (`sudo snap unset system experimental.parallel-instances`), rebooted, installed standard notes again.

There hasn't been an update of the snap yet, but I see that Standard Notes is still placing everything except the `.cache` folder into the revision specific directory.  Just like in my original post.  I strongly suspect Standard Notes will still lose all my notes on the next snap update.

TLDR: I've disabled the Parallel Installs feature but I believe the problem is still present in Standard Notes.

**Comment by pauldoo** [comment]
So after my last comment v3.6.4 was released on snap and the same problem occurred.  I lost all my notes and had to import from backup.  That confirms that the Parallel Installs feature is not the cause of the issue.

**Comment by baptiste-grob** [comment]
Hey @pauldoo, thanks a lot for raising this issue, creating a additional 'legacy' snap channel is certainly a good suggestion.

While IndexedDB and Local Storage data is indeed stored in the snap's revision folders it should transfer over when a next revision is installed. I updated to 3.6.4 on my own machine and all my data was still there.

I'm wondering if there could be something else going on here. Is parallel installs the only thing you've tried on your machine?

**Comment by pauldoo** [comment]
That’s the only snap configuration I’ve changed.  There are likely dozens of other differences on my system with respect to other packages I have and general system configuration.

Investigating this is going to be difficult if I cannot trigger an update on demand.  If you could publish another channel with a different version I can investigate this by:

1. Uninstall
2. Install older channel version
3. Upgrade to newer channel version
4. Repeat..

Could you do that, and perhaps tell me which logs I should look into?

**Comment by JaspalSuri** [comment]
> If you could publish another channel with a different version I can investigate this by:

For what it's worth, you can download prior versions of the app from the [releases page](https://github.com/standardnotes/desktop/releases/). 🙂

**Comment by baptiste-grob** [comment]
We've identified a case where when you update the snap while the app is running, changes you make after the update has completed will be lost. This only happens when you don't have an account, because otherwise you'll just pull the changes from the server.
This is something we're looking to address, and we'll be sure to update this thread once we make meaningful progress. Thank you for bringing this to our attention!
Though it doesn't quite match your experience, where _all_ your notes are wiped, correct?

> If you could publish another channel with a different version I can investigate this by:

We're currently running tests with a new version in `latest/edge`, however please do not use it unless you have a very recent manual backup handy! We'll let you know once we've got a potential fix published there.

**Comment by baptiste-grob** [comment]
Hey @pauldoo,

We've just released two versions (one on `latest/beta` and the other one on `latest/edge`) with a fix that moves user data to the snap's common folder like you suggested 🙂 If you want to try it out _and_ you're got a backup stored in a safe place, here's what you can do:

1. Change the snap's channel to `latest/beta`, then click install
1. Run the app and keep it open
1. Change the snap's channel to `latest/edge`, then click install
1. Once installation is complete, write something in your notes
1. Close and reopen Standard Notes

What you should notice is that while the move from `latest/stable` to `latest/beta` can still exhibit the bug you've described, moving from `latest/beta` to `latest/edge` shouldn't alter your data.
Since you've experienced a slightly different bug than what we've been able to reproduce (i-e you lose _all_ data between updates, not just some of it), we'd love to hear back from you if you are willing to try it!

Lastly, after this experimentation is done I'd recommend either staying an the beta channel until stable catches up to it (which should be a matter of days at this point), or removing the app and installing the stable version again.

**Comment by pauldoo** [comment]
Hi there,
I followed the steps you provided and the issue appears to be fixed.  Below are the exact details.

The only small issue I see is that the `"Standard Notes Backups"` folder is still stored by default in the revision specific folder, not `common`.  If someone needed to dig out an older backup they'd need to look into older revision folders.

## Detailed steps.

Here's what I did.

1. Backed up and quit.
2. Updated to `latest/beta` channel: `snap refresh standard-notes --channel=latest/beta` (this installed 3.6.5)
3. Opened standard notes.  All notes were gone (as expected).
4. I reconfigured (local only, no account, no password), and imported my backups.
5. Edited a note.
6. With standard notes still running, I switched to the `edge` channel: `snap refresh standard-notes --channel=latest/edge` (this installed 3.6.6)
7. Quit and reopened standard notes, all my content was preserved!

I see that there are files in the `common` directory now.

I've gone back to the `beta` channel for the time being.

**Comment by pauldoo** [comment]
I forgot to answer one of your specific questions:

> We've identified a case where when you update the snap while the app is running, changes you make after the update has completed will be lost. This only happens when you don't have an account, because otherwise you'll just pull the changes from the server.
> This is something we're looking to address, and we'll be sure to update this thread once we make meaningful progress. Thank you for bringing this to our attention!
> Though it doesn't quite match your experience, where all your notes are wiped, correct?

That is correct.  All my notes disappear, the app acts as if I've never used it before, and I get the initial setup popup (asks about password storage / keyring access IIRC).

Are there logs that would be useful to look at to debug this?

**Comment by JaspalSuri** [comment]
Hi @pauldoo,

I just wanted to let you know that @baptiste-grob will get back to you either tomorrow or early next week. 🙂 

I'm glad to hear that you were able to confirm that the latest version of the desktop app addresses the issue! 

> The only small issue I see is that the `Standard Notes Backups` folder is still stored by default in the revision specific folder, not `common`. If someone needed to dig out an older backup they'd need to look into older revision folders.

Thanks for pointing this out! I'll update our internal report by including this issue. 

> Are there logs that would be useful to look at to debug this?

While Baptiste would know more on this, the only logs that would be available would the console log in the dev tools (which wouldn't help here) and the logs in the directory that you referenced in the original post (which didn't contain anything at the time).

**Comment by baptiste-grob** [comment]
Happy to hear that fixed it for you! I think it's fair to say your issue was a variation of what we observed.

> The only small issue I see is that the `"Standard Notes Backups"` folder is still stored by default in the revision specific folder, not `common`. If someone needed to dig out an older backup they'd need to look into older revision folders.

Can you go to the top app menu, then Backup > Open Backups Location? Leftover backups cannot be deleted as a snap's previous revision folders are read-only, but it should have transferred them all to your documents folder.

**Comment by pauldoo** [comment]
> Can you go to the top app menu, then Backup > Open Backups Location and see where that takes you in your file explorer? 

This takes me to `/home/paul/snap/standard-notes/37/Standard Notes Backups`.

> Leftover backups cannot be deleted as a snap's previous revision folders are read-only, but it should have transferred them all to your documents folder.

There are no standard notes backups in my documents folder.

**Comment by JaspalSuri** [comment]
Hi @pauldoo, thank you for letting us know. I'll get in touch with @baptiste-grob on this and will get back to you.

**Comment by baptiste-grob** [comment]
Thanks for reporting back -- we'll have to make sure the backups migration is running properly for everyone then. In the meantime you can change your backups location manually if you'd like 👍
